### PR TITLE
Cleanup and refactoring

### LINKED ISSUE #517: Refactor the variant stuff
Right now it is in `atomic_judgments.sml`. That does not make any sense.

### PR DESCRIPTION
**Pull request by jonsterling**
I felt that the RedPRL codebase was becoming too sprawling and that individual modules were intimidating. As a first step, I've started reorganizing the modules that have to do with syntax; in some cases, I have moved code into its own modules (like "variance" "selector", etc.).

I've renamed some things too, as well as made better use of SML's datatype replication mechanism.

This resolves #517.

### CODE REVIEW

File: src/redprl.mlb [hunk 1]
```diff
@@ -73,7 +73,6 @@ in
 
 
     ann
-      "nonexhaustiveBind ignore"
```
Comment (favonia) [inline src/redprl.mlb]: Which version of MLton are you using? I believe this is a new feature introduced in 2017: (see the commit about their documentation https://github.com/MLton/mlton/commit/0908f6af4db86fd88461b358411a4488560838e2). If I understand it correctly, the `nonexhaustiveMatch` was broken into multiple annotations.

**Comment by favonia** [comment]
I am thinking that maybe `Selector` and `Accessor` can be put into the same files because they kind of have the same purpose, but separating them is also good.

File: src/redprl.mlb (hunk 1 above)
Comment (jonsterling) [reply src/redprl.mlb]: I guess my build is from 2016; I'm just using the one that you get in Ubuntu when you run `apt install mlton`.

Comment (favonia) [reply src/redprl.mlb]: I see. I was using a very recent version because I want https://github.com/MLton/mlton/pull/190.

**Comment by jonsterling** [comment]
@favonia 

> I am thinking that maybe Selector and Accessor can be put into the same files because they kind of have the same purpose, but separating them is also good.

Yes, I had that thought too... the only reason I didn't was because I didn't know what to name the file :laughing:

File: src/redprl.mlb (hunk 1 above)
Comment (jonsterling) [reply src/redprl.mlb]: Ah :smile:

**Comment by favonia** [comment]
Wait, I meant you should revert commit https://github.com/RedPRL/sml-redprl/pull/521/commits/dd5bdafc48adbc5f1ddbd6aa97ecdedae18b82ba unless it stops the old MLton from compiling.

**Comment by jonsterling** [comment]
@favonia I disagree; on the version of MLton that everyone except you likely has installed, it will cause a scary-looking warning during compilation. Unless it's absolutely necessary, I don't want our build to depend on versions of the compiler that are not easy to install using the standard packaging.

**Comment by jonsterling** [comment]
If you feel really strongly about it, we can add it back, though.

**Comment by favonia** [comment]
Alright. It's necessary for me to inspect possible bugs found by MLton, but I guess I can add it back on my local copy. FYI: Everyone needs to solve the PIE issue somehow, and Arch and its friends provide the updated `mlton`, while Debian/Ubuntu and their friends just patch the stable version of `mlton`. Hopefully after https://github.com/MLton/mlton/issues/147 this is no longer an issue.

**Comment by favonia** [comment]
Oops. I meant to say that the distros which use the updated MLton already has the new annotation design.

**Pull request merged**